### PR TITLE
Airbyte-ci: Add skippable connector test steps

### PR DESCRIPTION
**Pull request by bnchrch**
## Problem
As a connector developer I would like to be able to skip individual steps in the test pipeline

## Proposed Solution
```
airbyte-ci connectors --name=source-s3 test -x unit -x acceptance
```

## Implementation
Defining the Steps to be run as a list
```py
def get_test_steps(context: ConnectorContext) -> List[StepToRun]:
    """
    Get all the tests steps for a Python connector.
    """
    return [
        [StepToRun(id="build", step=BuildConnectorImages(context, LOCAL_BUILD_PLATFORM))],
        [StepToRun(
            id="unit",
            step=UnitTests(context),
            args=lambda results: {"connector_under_test": results["build"].output_artifact[LOCAL_BUILD_PLATFORM]},
            depends_on=["build"],
        )],
        [
            StepToRun(
                id="integration",
                step=IntegrationTests(context),
                args=lambda results: {"connector_under_test": results["build"].output_artifact[LOCAL_BUILD_PLATFORM]},
                depends_on=["build"],
            ),
            StepToRun(
                id="acceptance",
                step=AcceptanceTests(context, context.concurrent_cat),
                args=lambda results: {"connector_under_test_container": results["build"].output_artifact[LOCAL_BUILD_PLATFORM]},
                depends_on=["build"],
            ),
            StepToRun(id="check_base_image", step=CheckBaseImageIsUsed(context), depends_on=["build"]),
        ],
    ]

```

## Other minor changes
1. connector_secrets is now explicitly part of the ConnectorContext

### Why a list?
This will give us a type and structure we can manupilate prior to running, enabling features like
1. Skipping steps
2. Passing extra args
3. Eagerly failing tests on failing dependencies
4. Generating a test report prior to the tests completing

### Why a new class?
The new class `StepToRun` was added for two reasons
1. Its difficult to add this functionality to all steps without a much larger rewrite
2. The need for dynamic args when calling `run` still would require an addtional class / datamodel
3. This lets us keep all coordination logic inside the `run_steps.py` file

### CODE REVIEW

File: airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/context.py [hunk 1]
```diff
@@ -206,6 +206,11 @@ def docker_hub_password_secret(self) -> Optional[Secret]:
             return None
         return self.dagger_client.set_secret("docker_hub_password", self.docker_hub_password)
 
+    async def connector_secrets(self):
+        if self._connector_secrets is None:
+            self._connector_secrets = await secrets.get_connector_secrets(self)
+        return self._connector_secrets
+
```
Comment (alafanechere) [inline airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/context.py:213]: Is this related to your changes?

File: airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/test/commands.py [hunk 2]
```diff
@@ -30,27 +31,27 @@
...
 )
+@click.option(
+    '--skip-step',
+    '-x',
+    multiple=True,
+    type=str,
+    help="Skip a step by name. Can be used multiple times to skip multiple steps.",
+)
 @click.pass_context
 async def test(
```
Comment (alafanechere) [inline airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/test/commands.py:49]: Shall we add a `discover` subcommand or `--dry-run` flag to help users discover the available step names?

File: airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/context.py [hunk 3]
```diff
@@ -60,6 +61,7 @@ def __init__(
         s3_build_cache_access_key_id: Optional[str] = None,
         s3_build_cache_secret_key: Optional[str] = None,
         concurrent_cat: Optional[bool] = False,
+        run_step_options: RunStepOptions = RunStepOptions(),
```
Comment (alafanechere) [inline airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/context.py:64]: I think it would make sense to expose this parameter at the `PipelineContext` level for reusability. Wdyt?

File: airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/test/pipeline.py [hunk 4]
```diff
@@ -3,125 +3,67 @@
...
+    """
+    Compute the steps to run for a connector test pipeline.
+    """
 
-    A visual DAG can be found on the README.md file of the pipelines modules.
+    steps_to_run = get_test_steps(context)
 
-    Args:
-        context (ConnectorContext): The initialized connector context.
+    if not context.code_tests_only:
```
Comment (alafanechere) [inline airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/test/pipeline.py:49]: Shall we add a `code_test` boolean flag to `StepToRun` or `Step` to make this list dynamic?
I'm not against keeping it hardcoded but would prefer a constant list like:
```
STATIC_ANALYSIS_STEPS = [`MetadataValidation`, `VersionFollowsSemverCheck`, `VersionIncrementCheck`, `QaChecks`]
``` 
```

File: airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/test/pipeline.py [hunk 5]
```diff
@@ -3,125 +3,67 @@
...
-                        task_group.soonify(run_version_checks)(context),
-                        task_group.soonify(run_qa_checks)(context),
-                    ]
-            results = list(itertools.chain(*(task.value for task in tasks)))
+            result_dict = await run_steps(
+                runnables=steps_to_run,
+                options=context.run_step_options,
+            )
+
+            results = list(result_dict.values())
```
Comment (alafanechere) [inline airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/test/pipeline.py:66]: nit: I'm wondering if casting to list is mandatory here, does directly passing `result_dict.values()` to the ConnectorReport work?

File: airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/test/pipeline.py [hunk 6]
```diff
@@ -3,125 +3,67 @@
...
-                    tasks += [
-                        task_group.soonify(run_metadata_validation)(context),
-                        task_group.soonify(run_version_checks)(context),
-                        task_group.soonify(run_qa_checks)(context),
-                    ]
-            results = list(itertools.chain(*(task.value for task in tasks)))
+            result_dict = await run_steps(
+                runnables=steps_to_run,
+                options=context.run_step_options,
+            )
```
Comment (alafanechere) [inline airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/test/pipeline.py:64]: 💎

File: airbyte-ci/connectors/pipelines/pipelines/helpers/run_steps.py [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+            if steps.id in options.skip_steps:
+                main_logger.info(f"{indent * depth}- {steps.id} (skip)")
+            else:
+                main_logger.info(f"{indent * depth}- {steps.id}")
+
+
+async def run_steps(
+    runnables: STEP_TREE,
+    results: RESULTS_DICT = {},
+    options: RunStepOptions = RunStepOptions(),
```
Comment (alafanechere) [inline airbyte-ci/connectors/pipelines/pipelines/helpers/run_steps.py:151]: Shall we add a `concurrency` attribute to `RunStepOptions` which would lead to the definition of a semaphore in this function? We currently control concurrency at the connector level but if `steps_to_run` is huge we might like to cap the concurrency.

File: airbyte-ci/connectors/pipelines/pipelines/helpers/run_steps.py [hunk 8]
```diff
@@ -0,0 +1,233 @@
...
+            main_logger.info(f"{indent * depth}- {steps[0].id} (parallel)")
+            _log_step_tree(steps, options, depth + 1)
+        else:
+            if steps.id in options.skip_steps:
+                main_logger.info(f"{indent * depth}- {steps.id} (skip)")
+            else:
+                main_logger.info(f"{indent * depth}- {steps.id}")
+
+
+async def run_steps(
```
Comment (alafanechere) [inline airbyte-ci/connectors/pipelines/pipelines/helpers/run_steps.py:148]: 💎 💎 💎 💎 💎 💎 💎 💎 💎

File: airbyte-ci/connectors/pipelines/tests/test_helpers/test_run_steps.py [hunk 9]
```diff
@@ -0,0 +1,238 @@
...
+            [StepToRun(id="step3", step=TestStep(test_context))],
+            [StepToRun(id="step4", step=TestStep(test_context))],
+        ],
+        {
+            "step1": StepStatus.SUCCESS,
+            "step2": StepStatus.SUCCESS,
+            "step3": StepStatus.SUCCESS,
+            "step4": StepStatus.SUCCESS
+        },
+        RunStepOptions(fail_fast=True)
```
Comment (alafanechere) [inline airbyte-ci/connectors/pipelines/tests/test_helpers/test_run_steps.py:35]:
```suggestion
       id="Something meaningful describing the test case"
        [
            [StepToRun(id="step1", step=TestStep(test_context))],
            [StepToRun(id="step2", step=TestStep(test_context))],
            [StepToRun(id="step3", step=TestStep(test_context))],
            [StepToRun(id="step4", step=TestStep(test_context))],
        ],
        {
            "step1": StepStatus.SUCCESS,
            "step2": StepStatus.SUCCESS,
            "step3": StepStatus.SUCCESS,
            "step4": StepStatus.SUCCESS
        },
        RunStepOptions(fail_fast=True)
```
Would be helpful to grasp what's tested or not

File: airbyte-ci/connectors/pipelines/pipelines/helpers/run_steps.py [hunk 10]
```diff
@@ -0,0 +1,233 @@
...
+
+
+async def evaluate_run_args(args: ARGS_TYPE, results: RESULTS_DICT) -> Dict:
+    """
+    Evaluate the args of a StepToRun using the results of previous steps.
+    """
+    if inspect.iscoroutinefunction(args):
+        return await args(results)
+    elif callable(args):
+        return args(results)
```
Comment (alafanechere) [inline airbyte-ci/connectors/pipelines/pipelines/helpers/run_steps.py:54]: According to your type hints `args` can be a `dict` but here you consider its always a `callable`

File: airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/test/steps/python_connectors.py [hunk 11]
```diff
@@ -197,48 +163,32 @@ class IntegrationTests(PytestStep):
...
+    Get all the tests steps for a Python connector.
     """
-    return [await CodeFormatChecks(context).run()]
+    return [
+        [StepToRun(id="build", step=BuildConnectorImages(context, LOCAL_BUILD_PLATFORM))],
+        [StepToRun(
+            id="unit",
+            step=UnitTests(context),
+            args=lambda results: {"connector_under_test": results["build"].output_artifact[LOCAL_BUILD_PLATFORM]},
+            depends_on=["build"],
```
Comment (alafanechere) [inline airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/test/steps/python_connectors.py:176]: If I'm not mistaken the dependency is always implicitly defined by the fact that a Step args are from a dependency `output_artifact`. Could we infer the dependency tree dynamically from the args then?
I know explicit is better than implicit 😄 , so feel free to reject this suggestion

File: airbyte-ci/connectors/pipelines/tests/test_helpers/test_run_steps.py
Comment (alafanechere) [inline airbyte-ci/connectors/pipelines/tests/test_helpers/test_run_steps.py:1]: Could you add tests that check that argument passing between steps works?

File: airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/test/steps/java_connectors.py [hunk 12]
```diff
@@ -64,64 +64,91 @@ class UnitTests(GradleTask):
...
+
+def get_test_steps(context: ConnectorContext) -> List[StepToRun]:
+    """
+    Get all the tests steps for a Java connector.
+    """
 
-    return step_results + [soon_unit_tests_result.value] + soon_docker_build_dependent_steps_results.value
+    steps = [
+        [StepToRun(id="build_tar", step=BuildConnectorDistributionTar(context))],
+        [StepToRun(id="unit", step=UnitTests(context))],
```
Comment (alafanechere) [inline airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/test/steps/java_connectors.py:136]:
```suggestion
        [StepToRun(id="build_tar", step=BuildConnectorDistributionTar(context))],
        [StepToRun(id="unit", step=UnitTests(context), depends_on=["build_tar"])],
```
Even if there are no functional requirement to make unit test depend on build tar I'd prefer to express a dependency here as `build_tar` will warm the unit test gradle cache.

File: airbyte-ci/connectors/pipelines/pipelines/helpers/run_steps.py [hunk 13]
```diff
@@ -0,0 +1,233 @@
...
+class StepToRun:
+    """
+    A class to wrap a Step with its id and args.
+
+    Used to coordinate the execution of multiple steps inside a pipeline.
+    """
+    id: str
+    step: Step
+    args: ARGS_TYPE = field(default_factory=dict)
+    depends_on: List[str] = field(default_factory=list)
```
Comment (alafanechere) [inline airbyte-ci/connectors/pipelines/pipelines/helpers/run_steps.py:42]: I would find it more correct and robust if the `depends_on` type was a list of `StepToRun` instances

File: airbyte-ci/connectors/pipelines/pipelines/helpers/run_steps.py [hunk 14]
```diff
@@ -0,0 +1,233 @@
...
+@dataclass
+class RunStepOptions:
+    """Options for the run_step function."""
+    fail_fast: bool = True
+    skip_steps: List[str] = field(default_factory=list)
+    log_step_tree: bool = True
+
+
+@dataclass(frozen=True)
+class StepToRun:
```
Comment (alafanechere) [inline airbyte-ci/connectors/pipelines/pipelines/helpers/run_steps.py:33]: Just opening the discussion here.
Could we rename `StepToRun` to `Step`
And `Step` to `Operation`?

A `Operation.__init__` would take a `Step` instance and would access parameters from `Step.args`.

`Step` instances would be "singleton"
`Operation` instance would not be singleton: an operation can be  re-used in multiple step.
And I think `Operation` might be worth being functions as they'll be stateless.

**Review by alafanechere — COMMENTED** [review]
I LOVE the direction where this is going.

File: airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/test/commands.py [hunk 15]
```diff
@@ -30,27 +31,27 @@
...
     type=bool,
     is_flag=True,
 )
+@click.option(
+    '--skip-step',
+    '-x',
+    multiple=True,
+    type=str,
+    help="Skip a step by name. Can be used multiple times to skip multiple steps.",
+)
```
[the remark this answers is not in the document]
Comment (bnchrch) [reply airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/test/commands.py:47]: Not much at all! I can do that in another PR

File: airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/test/commands.py [hunk 16]
```diff
@@ -30,27 +31,27 @@
...
     help="When enabled, the CAT tests will run concurrently. Be careful about rate limits",
     default=False,
     type=bool,
     is_flag=True,
 )
+@click.option(
+    '--skip-step',
+    '-x',
+    multiple=True,
+    type=str,
```
[the remark this answers is not in the document]
Comment (bnchrch) [reply airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/test/commands.py:45]: 🤔 Good idea. Let me add that

File: airbyte-ci/connectors/pipelines/pipelines/helpers/run_steps.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (bnchrch) [reply airbyte-ci/connectors/pipelines/pipelines/helpers/run_steps.py:42]: I thought about that for awhile

However we have a few generic steps that are reused. Like PytestStep, SimpleDockerStep and others

File: airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/context.py (hunk 1 above)
Comment (bnchrch) [reply airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/context.py:213]: It is! We were mounting connector secrets in between steps. So I hoisted that logic into the context so when you request `connector_secrets` it mounts and gets them if it hasnt already

File: airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/test/commands.py (hunk 2 above)
Comment (bnchrch) [reply airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/test/commands.py:49]: Once I make this change you suggested
https://github.com/airbytehq/airbyte/pull/32188#discussion_r1384937430

Then --help should allow them to discover the step names

File: airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/test/pipeline.py (hunk 5 above)
Comment (bnchrch) [reply airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/test/pipeline.py:66]: Good call out, I think your right, let me check

File: airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/test/pipeline.py (hunk 4 above)
Comment (bnchrch) [reply airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/test/pipeline.py:49]: I was thinking of adding a `tags` field to the `StepToRun` class in another PR for this

so we could mark steps as `slow` or `code_test` then the user could run

```
airbyte-ci connector test -x slow --only code_test
```

wdyt?

File: airbyte-ci/connectors/pipelines/tests/test_helpers/test_run_steps.py (hunk 9 above)
Comment (bnchrch) [reply airbyte-ci/connectors/pipelines/tests/test_helpers/test_run_steps.py:35]: Good call!

File: airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/test/pipeline.py (hunk 4 above)
Comment (alafanechere) [reply airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/test/pipeline.py:49]: 👍

File: airbyte-ci/connectors/pipelines/pipelines/helpers/run_steps.py (hunk 10 above)
Comment (bnchrch) [reply airbyte-ci/connectors/pipelines/pipelines/helpers/run_steps.py:54]: Hmm maybe I dont follow but I think the type is correct 🤔 

```
ARGS_TYPE = Union[Dict, Callable[[RESULTS_DICT], Dict]]
```

Means args is either a `dict` or a function that returns a `dict`

File: airbyte-ci/connectors/pipelines/pipelines/helpers/run_steps.py (hunk 13 above)
Comment (bnchrch) [reply airbyte-ci/connectors/pipelines/pipelines/helpers/run_steps.py:42]: Im not full against that but why is that?

Comment (alafanechere) [reply airbyte-ci/connectors/pipelines/pipelines/helpers/run_steps.py:42]: A child could know its parents if `depends_on` was a `StepToRun`

Comment (alafanechere) [reply airbyte-ci/connectors/pipelines/pipelines/helpers/run_steps.py:42]: And it limits the  id typos as we'd pass around a variable

File: airbyte-ci/connectors/pipelines/pipelines/helpers/run_steps.py (hunk 14 above)
Comment (bnchrch) [reply airbyte-ci/connectors/pipelines/pipelines/helpers/run_steps.py:33]: Im all for finding a way to cut down the awkwardness (and layers) of these abstraction.


Hearing works like "singleton", "would be" and "I think" is telling me that this might be better tackled on its own so we can just wrestle with restructuring the taxonomy and not restructuring the taxonomy AND adding new functionality.

WDYT?

File: airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/context.py (hunk 1 above)
Comment (alafanechere) [reply airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/context.py:213]: Could we rename the function to `get_connector_secrets` then? Or would it make sense to make `secrets.get_connector_secrets` a method of the context class?

File: airbyte-ci/connectors/pipelines/pipelines/helpers/run_steps.py (hunk 7 above)
Comment (alafanechere) [reply airbyte-ci/connectors/pipelines/pipelines/helpers/run_steps.py:151]: Yes, you'd instantiate a Semaphore with the maximum capacity in the `RunStepOptions`
and wrap the task group on which you'd like to control the concurrency with  `with options.semaphore`

File: airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/context.py (hunk 1 above)
Comment (bnchrch) [reply airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/context.py:216]: Im torn on this

On one hand it belongs in `context` given how coupled it is to a `ConnectorContext`

On the other hand connector secrets is a foundational concept that deserves a separate file to hold onto the logic.

I think we gain more from it being separate.

WDYT?

File: airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/context.py (hunk 3 above)
Comment (bnchrch) [reply airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/context.py:64]: We do! Im just passing it through to the pipeline context here

File: airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/test/steps/python_connectors.py (hunk 11 above)
Comment (bnchrch) [reply airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/test/steps/python_connectors.py]: Thats fair! but there sometimes output that may not be an artifact, thus theres no implicit relationship either.

e.g. we have a step called `PushToDockerHub` and a step called `VerifyOnDockerHub`

**Comment by bnchrch** [comment]
@alafanechere This is ready for review again!

Since the last review
1. Updated to latest master
2. Added support for `--help` listing skippable steps for `-x`
3. Made `StepToRun` ids less vulnerable to typos
4. Added checks for args being one of callable, awaitable or dict
4. Updated tests to have descriptions
5. Added test cases for skip steps, fast failures, concurrency, invalid args and more
6. Added configurable concurrency support


I also left a few discussions open around `depends_on` and `get_connector_secrets

File: airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/context.py (hunk 1 above)
Comment (alafanechere) [reply airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/context.py:216]: I suggest renaming it to `get_connector_secrets` as it:
* Downloads the connector secrets
* Assign the secrets to `self._connector_secrets`.

And we rename `self._connector_secrets` to `self.connector_secrets` to make it public.

I'd be confused to have a method named `connector_secrets` acting as a getter on `self._connector_secrets`

File: airbyte-ci/connectors/pipelines/pipelines/consts.py [hunk 17]
```diff
@@ -89,3 +89,26 @@ class INTERNAL_TOOL_PATHS(str, Enum):
 
 
 DAGGER_WRAP_ENV_VAR_NAME = "_DAGGER_WRAP_APPLIED"
+
+
+class CONNECTOR_TEST_STEP_ID(str, Enum):
```
Comment (alafanechere) [inline airbyte-ci/connectors/pipelines/pipelines/consts.py:94]: What do you think about moving this enum to `connectors/consts.py` as it's connector-specific?

**Review by alafanechere — CHANGES_REQUESTED** [review]
I think there's a bug in the input validation:
`airbyte-ci connectors --name=source-faker test -x=acceptance`:

```
┃ Error: Invalid value for '--skip-step' / '-x': '=acceptance' is not one of <CONNECTOR_TEST_STEP_ID.ACCEPTANCE: 'acceptance'>, <CONNECTOR_TEST_STEP_ID.BUILD_NORMALIZATION: 'build_normalization'>, <CONNECTOR_TEST_STEP_ID.BUILD_TAR: 'build_tar'>, <CONNECTOR_TEST_STEP_ID.BUILD: 'build'>, <CONNECTOR_TEST_S
┃ TEP_ID.CHECK_BASE_IMAGE: 'check_base_image'>, <CONNECTOR_TEST_STEP_ID.INTEGRATION: 'integration'>, <CONNECTOR_TEST_STEP_ID.METADATA_VALIDATION: 'metadata_validation'>, <CONNECTOR_TEST_STEP_ID.QA_CHECKS: 'qa_checks'>, <CONNECTOR_TEST_STEP_ID.UNIT: 'unit'>, <CONNECTOR_TEST_STEP_ID.VERSION_FOLLOW_CHECK:
┃ 'version_follow_check'>, <CONNECTOR_TEST_STEP_ID.VERSION_INC_CHECK: 'version_inc_check'>, <CONNECTOR_TEST_STEP_ID.TEST_ORCHESTRATOR: 'test_orchestrator'>, <CONNECTOR_TEST_STEP_ID.DEPLOY_ORCHESTRATOR: 'deploy_orchestrator'>.
```

**Comment by bnchrch** [comment]
I think thats just a quirk of click.choice

We have the same issue with the `-a` flag in build

File: airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/context.py (hunk 1 above)
Comment (bnchrch) [reply airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/context.py:216]: Sounds good! Though, lets leave the internal field with the underscore. We want devs to always reference the getter and feel bad otherwise

**Comment by bnchrch** [comment]
@alafanechere Ready for your review again :)

**Comment by alafanechere** [comment]
@bnchrch I figured that `=` can be used with long options `--` but with short one a space is expected... Might be a convention I was not aware of.
I changed the `Choice` a bit to have a cleaner output in case the passed `-x` value is invalid

**Review by alafanechere — APPROVED** [review]
Ship it 🎉 
I think I fixed the test suite + conflicts.
I manually tested `connectors test` , everything is working as expected so far!
